core/dns.go

```diff
@@ -32,23 +34,35 @@ var (
 		IP:   []byte{192, 168, 0, 0},
 		Mask: []byte{255, 255, 0, 0},
 	}
+
+	// RFC5735
+	// 127.0.0.0/8
+	rfc5735_127 = net.IPNet{
+		IP:   []byte{127, 0, 0, 0},
+		Mask: []byte{255, 0, 0, 0},
+	}
 )
 
 // DNSResolverImpl represents a client that talks to an external resolver
 type DNSResolverImpl struct {
-	DNSClient *dns.Client
-	Servers   []string
+	DNSClient              *dns.Client
+	Servers                []string
+	allowLoopbackAddresses bool
 }
 
 // NewDNSResolverImpl constructs a new DNS resolver object that utilizes the
 // provided list of DNS servers for resolution.
-func NewDNSResolverImpl(dialTimeout time.Duration, servers []string) *DNSResolverImpl {
+func NewDNSResolverImpl(dialTimeout time.Duration, servers []string, allowLoopback bool) *DNSResolverImpl {
```

Comment:
Another alternative is its constructor taking a function that has the same type as isPrivateV4 and we have a test vs prod one defined in the public API.